File: Demos/Mandelbrot/Algorithm.cs
using System;

namespace Mandelbrot
{
	/// <summary>
	/// Provides two different ways of calculating the Mandelbrot set. 
	/// One uses Complex arithmetic, the other only real numbers.
	/// </summary>
	public class Algorithm
	{
		private int m_MaxIterations;

		public Algorithm()
		{
			m_MaxIterations = 128;
		}

		public bool UseComplexNumbers = false;

		public int MaxIterations
		{
			get
			{
				return m_MaxIterations;
			}

			set
			{
				m_MaxIterations = value;
			}
		}

		/// <summary>
		/// Calculates the number of Mandelbrot iterations. 
		/// </summary>
		/// <param name="x">real starting value</param>
		/// <param name="y">complex starting value</param>
		/// <returns></returns>
		public int Evaluate(double x, double y)
		{
			if(UseComplexNumbers)
				return EvaluateUsingComplexNumbers(x, y);
			else
				return EvaluateUsingDoubles(x, y);
		}

		/// <summary>
		/// Calculates the number of Mandelbrot iterations using complex numbers.
		/// </summary>
		/// <param name="x">real starting value</param>
		/// <param name="y">complex starting value</param>
		/// <returns></returns>
		int EvaluateUsingComplexNumbers(double x, double y)
		{
			int iterations = 0;
			Complex z = new Complex(x, y);
			Complex z0 = new Complex(z);
			while(iterations < m_MaxIterations && z.ModSquared < 4)
			{
				iterations++;
				// this is the Mandelbrot algorithm in complex form.
				z = z * z + z0;
			}

			return iterations;
		}

		/// <summary>
		/// Calculates the number of Mandelbrot iterations using only doubles
		/// </summary>
		/// <param name="x">real starting value</param>
		/// <param name="y">complex starting value</param>
		/// <returns></returns>
		int EvaluateUsingDoubles(double x, double y)
		{
			double x0 = x;
			double y0 = y;
			double ysquared = y0 * y0;
			double xsquared = x0 * x0;
			int iterations = 0;
			while( (iterations < m_MaxIterations) && (x0*x0 + y0*y0 < 4) )
			{
				// this is the Mandelbrot algorithm using doubles.
				ysquared = y0 * y0;
				xsquared = x0 * x0;
				y0 = 2 * y0 * x0 + y;
				x0 = xsquared - ysquared + x;
				iterations++;
			}

			return iterations;
		}
	}
}

File: Demos/MandelbrotVB/Algorithm.vb
Imports System

Namespace Mandelbrot

	Public Class Algorithm
		Private m_MaxIterations As Integer

		Public Sub New()
			m_MaxIterations = 128
		End Sub
		Public UseComplexNumbers As Boolean = False

		Public Property MaxIterations() As Integer
			Get
				Return m_MaxIterations
			End Get
			Set
				m_MaxIterations = value
			End Set
		End Property

		Public Function Evaluate(ByVal x As Double, ByVal y As Double) As Integer
			If UseComplexNumbers Then
				Return EvaluateUsingComplexNumbers(x, y)
			Else
				Return EvaluateUsingDoubles(x, y)
			End If
		End Function

		Function EvaluateUsingComplexNumbers(ByVal x As Double, ByVal y As Double) As Integer
			Dim iterations As Integer = 0
			Dim z As Complex = New Complex(x, y)
			Dim z0 As Complex = New Complex(z)
            While iterations < m_MaxIterations AndAlso z.ModSquared < 4
                z = Complex.Multiply(z, z)
                z = Complex.Add(z, z0)
                iterations = iterations + 1
            End While
            Return iterations
		End Function

		Function EvaluateUsingDoubles(ByVal x As Double, ByVal y As Double) As Integer
			Dim x0 As Double = x
			Dim y0 As Double = y
			Dim ysquared As Double = y0 * y0
			Dim xsquared As Double = x0 * x0
			Dim iterations As Integer = 0
			While (iterations < m_MaxIterations) AndAlso (x0 * x0 + y0 * y0 < 4)
				ysquared = y0 * y0
				xsquared = x0 * x0
				y0 = 2 * y0 * x0 + y
                x0 = xsquared - ysquared + x
                iterations = iterations + 1
            End While
            Return iterations
		End Function
	End Class 
End Namespace
